config/crd/snapshot.storage.k8s.io_volumesnapshots.yaml

```diff
@@ -0,0 +1,130 @@
+        underlying storage system, it is bound to a corresponding VolumeSnapshotContent.
+      properties:
+        apiVersion:
+          description: 'APIVersion defines the versioned schema of this representation
+            of an object. Servers should convert recognized schemas to the latest
+            internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#resources'
+          type: string
+        kind:
+          description: 'Kind is a string value representing the REST resource this
+            object represents. Servers may infer this from the endpoint the client
+            submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds'
+          type: string
+        metadata:
+          description: 'Standard object''s metadata. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata'
+          type: object
+        spec:
+          description: 'spec defines the desired characteristics of a snapshot requested
+            by a user. More info: https:s//kubernetes.io/docs/concepts/storage/volume-snapshots#volumesnapshots'
+          properties:
+            source:
+              description: source has the information about where a snapshot should
+                be created from. Currently PersistentVolumeClaim is the only supported
+                source type. If specified, and VolumeSnapshotContentName is not specified
+                (i.e., nil), snapshot will be dynamically created from the given source.
+                If not specified, user can statically bind a VolumeSnapshot to a VolumeSnapshotContent
+                by specifying VolumeSnapshotContentName.
+              properties:
+                apiGroup:
+                  description: APIGroup is the group for the resource being referenced.
+                    If APIGroup is not specified, the specified Kind must be in the
+                    core API group. For any other third-party types, APIGroup is required.
+                  type: string
+                kind:
+                  description: Kind is the type of resource being referenced
+                  type: string
+                name:
+                  description: Name is the name of resource being referenced
+                  type: string
+              required:
+              - kind
+              - name
+              type: object
+            volumeSnapshotClassName:
+              description: 'volumeSnapshotClassName is name of the VolumeSnapshotClass
+                requested by the VolumeSnapshot. If not specified, the default snapshot
+                class will be used if there exists one. More info: https://kubernetes.io/docs/concepts/storage/volume-snapshot-classes'
+              type: string
+            volumeSnapshotContentName:
```

Comment:
I was discussing an idea with with @jingxu97 and @saad-ali .  Can we make VolumeSnapshotContentName in the status?  The reason why it's in spec for PVC Is because we let users choose PV.  But it doesn't seem to be the case for snapshots.  Snapshots requires both user and admin to set the pointers, but in that case, do we really need the user to set it? Can we just trust the Content?  The only scenario I can think of is if there's multiple contents pointing to the same snapshot. But then if we still keep the field in snapshot, but put it in status, then I think that protects that case.

Reply:
There are couple of considerations worth clarifying if we'd take the suggested approach to put SnapshotContentName into status.
1. the current implementation distinguishes pre-existing snapshot content binding v.s. dynamic snapshot creation by checking whether a Snapshot object has the SnapshotContentName specified or not. If moved to status, how should that happen? One thing I can think of is to check whether Snapshot.Spec.Source is specified, but what if users actually want to have the source specified in pre-existing binding case? i.e., user knows from which PVC the snapshot was taken from.
2. Whether SnapshotContentName should be a status field or a spec field is a bit blur here. Having it in status field somehow means the bi-directional binding no longer holds for Snapshot and SnapshotContent, rather its a one-way binding from SnapshotContent ->  Snapshot given status fields can not guarantee the state of truth. I am OK with this model, just thinking what should be the controller logic if the binding has been broken, i.e., SnapshotContent was deleted from system, would controller relies on the status field of Snapshot to check?

Reply:
1. Yes, it could use nil Source to determine dynamic snapshot or not.  I'm not sure about a solution where you want to track the associated PVC. Is that information that can go into Content?
2. Yes it's a one-way binding from Content->Snapshot. It means that the user doesn't have a say into which snapshot they get; only the admin gets to decide.  I think that is an ok model.  What controller case are you thinking of where Content is deleted? What would controller try to do with the Snapshot?

Reply:
1. Source information has been removed from content, and of course we could add it back, also contents are non-namespaced and thus users wont have access.
2. Yes, what would controller behave when SnapshotContent is gone. It would need to evaluate based of the SnapshotContentName field in status which is not suggested.
From user's perspective, having it in spec could be beneficial because it's less ambiguity to distinguish dynamic v.s. pre-existing binding, follows the similar pattern of PV/PVC. IMO, the key here would be whether SnapshotContentName should be a desired state from user perspective or rather a status to them. My personal take would still make it user controlled field which allows them to ask for a specific SnapshotContent. If I am a user, in the pre-existing binding case, I'd prefer to know the content in advance so to make sure I am using the right one.

Reply:
1. How important is it to the user to know which PVC the Snapshot is for in the pre-created case?  
2. Why is absence of SnapshotContent not an error?

Setting ContentName is not exactly the same pattern as PVC/PV, because it requires Content-side to also be set. So the user input doesn't really matter in this case.

Reply:
Another problem with setting both contentName and source in snapshot object for the precreated case, is what happens if it's wrong?  What if the content is actually not from the PVC that the user specified in the source?  Could that cause problems especially if there are any higher controllers acting on this information?